### ISSUE TITLE
Third Purity value

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#7013]
Instead of being pure or impure, it should be pure, impure, or control-pure

**Comment by magnus-madsen** [comment]
@JonathanStarup This should actually be possible now -- because we monomorph with the concrete effect set now. If it only contains IO then it is impure. If it contains any user-defined effect then it is control impure.

**Comment by JonathanStarup** [comment]
A complication - the output of monomorph is only ground, its still a formula. This means that the `Type -> Purity` function is not trivial

**Comment by magnus-madsen** [comment]
It is ground, i.e. no variables, right? So then you just write an evaluator for set expressions? :)

**Comment by JonathanStarup** [comment]
Yes, but they can evaluate to Univ, so it need a wrapper of the set. And its cofinite, so it needs
```
enum SymSet {
   case Top, TopMinus(set), Set(set)
}
```
I just wanted to be sure that this is what you intended to happen

**Comment by magnus-madsen** [comment]
> Yes, but they can evaluate to Univ, so it needs a wrapper of the set. And its cofinite, so it needs
> 
> ```
> enum SymSet {
>    case Top, TopMinus(set), Set(set)
> }
> ```
> 
> I just wanted to be sure that this is what you intended to happen

I understand - in theory - that can happen. In practice, I guess we still don't really know if such a type could occur.

But what about, when you encounter Univ, you replace it with `root.effects.values.toSet`?

**Comment by JonathanStarup** [comment]
> you replace it with root.effects.values.toSet?

We could yes, but I think it'd nice not to make that assumption if we don't have to, and I think its theoretically faster not to do it. If the code for it is too complicated I agree we should do it. 

> In practice, I guess we still don't really know if such a type could occur

It can for sub-formulas certainly

**Comment by magnus-madsen** [comment]
But more importantly the complexity does not buy us anything.
